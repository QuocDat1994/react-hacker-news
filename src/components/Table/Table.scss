.table {
  width: 100%;
  border-collapse: collapse;

  thead {
    tr {
      th {
        font-weight: 500;
        position: relative;
        cursor: pointer;

        &:hover {
          background: #eefaf4;
          transition: 0.3s ease-in-out;
        }

        &.sortDown {
          svg {
            top: 45%;
          }
        }

        &.sortUp {
          svg {
            top: 25%;
          }
        }

        svg {
          position: absolute;
          margin-left: 5px;
        }

        &:last-child {
          cursor: none;
          pointer-events: none;
        }
      }
    }
  }

  tbody {
    tr {
      transition: 0.3s;

      &:hover {
        background: #eefaf4;
      }

      &:last-child {
        td {
          padding-bottom: 10px;
        }
      }
    }
  }

  td,
  th {
    border-bottom: 1px solid rgba($color: #000000, $alpha: 0.1);
    height: 70px;
    text-align: center;
    padding: 0 20px;

    &:first-child,
    &:nth-child(2) {
      text-align: left;
    }

    &:first-child {
      padding-left: 40px;
    }

    &:last-child {
      padding-right: 40px;
    }
  }

  td {
    .title {
      font-size: 18px;
      font-weight: 500;
      color: #74d39c;
      text-decoration: none;

      &:hover {
        color: darken(#74d39c, 30%);
        transition: 0.3s ease-in-out;
      }
    }

    .author {
      font-size: 14px;
    }

    button {
      height: 50%;
      width: 100px;
      color: white;
      background: #74d39c;
      border: none;
      border-radius: 100px;
      cursor: pointer;

      &:hover {
        background: darken(#74d39c, 30%);
        transition: 0.3s ease-in-out;
      }
    }
  }
}
