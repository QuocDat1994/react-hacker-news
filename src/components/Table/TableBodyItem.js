import React from "react";
import moment from "moment";

export default class TableBodyItem extends React.Component {
  handleClick = () => {
    this.props.onClick(this.props.data.objectID);
  };

  render() {
    const { data } = this.props;
    return (
      <tr>
        <td>
          <a className="title" href={data.url} target="_blank">
            {data.title}
          </a>
          <div className="author">by {data.author}</div>
        </td>
        <td>{moment(data.created_at).format("MMMM Do YYYY")}</td>
        <td>{data.num_comments}</td>
        <td>{data.points}</td>
        <td>
          <button onClick={this.handleClick}>Dismiss</button>
        </td>
      </tr>
    );
  }
}
