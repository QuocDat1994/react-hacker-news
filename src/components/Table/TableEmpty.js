import React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faFrown } from "@fortawesome/free-solid-svg-icons";

const TableEmpty = () => {
  return (
    <div className="table table--empty">
      <FontAwesomeIcon icon={faFrown} />
      <div>No Results</div>
      <p>
        There is no posts that matches your search. <br />
        Please try different search terms.
      </p>
    </div>
  );
};

export default TableEmpty;
