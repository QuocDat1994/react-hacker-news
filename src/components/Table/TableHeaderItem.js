import React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faSortUp, faSortDown } from "@fortawesome/free-solid-svg-icons";

export default class TableHeaderItem extends React.Component {
  handleClick = () => {
    this.props.onClick(this.props.text.toUpperCase());
  };

  getSortClass = () => {
    const { text, sortKey } = this.props;

    if (sortKey === text.toUpperCase()) {
      return "sortUp";
    } else if (sortKey === `${text.toUpperCase()}_DESC`) {
      return "sortDown";
    }
    return "";
  };

  render() {
    return (
      <th onClick={this.handleClick} className={this.getSortClass}>
        {this.props.text}
        {this.getSortClass() === "sortUp" && (
          <FontAwesomeIcon icon={faSortUp} />
        )}
        {this.getSortClass() === "sortDown" && (
          <FontAwesomeIcon icon={faSortDown} />
        )}
      </th>
    );
  }
}
