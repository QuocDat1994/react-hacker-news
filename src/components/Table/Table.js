import React from "react";
import "./Table.scss";

import TableHeaderItem from "./TableHeaderItem";
import TableBodyItem from "./TableBodyItem";

export default class Table extends React.Component {
  handleSort = key => {
    this.props.onSortChange(key);
  };

  handleDismiss = id => {
    this.props.onDismiss(id);
  };

  render() {
    return (
      <table className="table">
        <thead>
          <tr>
            <TableHeaderItem
              text="Title"
              onClick={this.handleSort}
              sortKey={this.props.sortKey}
            />
            <TableHeaderItem
              text="Date"
              onClick={this.handleSort}
              sortKey={this.props.sortKey}
            />
            <TableHeaderItem
              text="Comments"
              onClick={this.handleSort}
              sortKey={this.props.sortKey}
            />
            <TableHeaderItem
              text="Points"
              onClick={this.handleSort}
              sortKey={this.props.sortKey}
            />
            <th />
          </tr>
        </thead>
        <tbody>
          {this.props.list.map(data => (
            <TableBodyItem
              key={data.objectID}
              data={data}
              onClick={this.handleDismiss}
            />
          ))}
        </tbody>
      </table>
    );
  }
}
