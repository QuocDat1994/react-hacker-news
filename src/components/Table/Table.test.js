import React from "react";
import Enzyme, { shallow, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import Table from "./Table";

Enzyme.configure({ adapter: new Adapter() });

describe("Table", () => {
  const data = [
    {
      objectID: "16582136",
      created_at: "2018-03-14T03:50:30.000Z",
      title: "Stephen Hawking has died",
      url: "http://www.bbc.com/news/uk-43396008",
      author: "Cogito",
      points: 6015
    },
    {
      objectID: "11116274",
      created_at: "2016-02-17T08:38:37.000Z",
      title: "A Message to Our Customers",
      url: "http://www.apple.com/customer-letter/",
      author: "epaga",
      points: 5771
    },
    {
      objectID: "3078128",
      created_at: "2011-10-05T23:42:23.000Z",
      title: "Steve Jobs has passed away.",
      url: "http://www.apple.com/stevejobs/",
      author: "patricktomas",
      points: 4271
    }
  ];

  it("should render table with data", () => {
    const wrapper = shallow(<Table list={data} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should render table with data", () => {
    const wrapper = mount(<Table list={data} />);
    expect(wrapper.find("tbody tr").length).toBe(3);
  });

  it("should render empty table", () => {
    const wrapper = shallow(<Table list={[]} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should render empty table", () => {
    const wrapper = mount(<Table list={[]} />);
    expect(wrapper.find("tbody tr").length).toBe(0);
  });
});
