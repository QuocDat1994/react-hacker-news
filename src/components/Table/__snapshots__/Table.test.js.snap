// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should render empty table 1`] = `ShallowWrapper {}`;

exports[`Table should render table with data 1`] = `ShallowWrapper {}`;

exports[`Table should render without error 1`] = `ShallowWrapper {}`;
